engine/uState.ml

```diff
@@ -310,10 +446,13 @@ let process_universe_constraints uctx cstrs =
     else sort_inconsistency Eq l r
   in
   let unify_universes cst local =
-    let cst = nf_constraint cst in
+    let cst = nf_constraint local.local_sorts cst in
     if UnivProblem.is_trivial cst then local
     else match cst with
     | ULe (l, r) ->
+      let local = unify_quality univs Reduction.CUMUL l r local in
```

Comment:
The compat issue is probably from here, if `l = Prop` and `r` is a variable quality this forces it to be Prop.

Reply:
But we have to take a choice here, since it must rule out SProp. We could have a finer mechanism than just union-find but then we need a form of constraint system over qualities.

Reply:
I am still not fully convinced that making Prop a quality separate from Type can be made to work with such problems. Perhaps we just have to accept those problems.

Reply:
I don't see any foundational problem, and furthermore we need to keep Prop and Type qualities separated for impredicativity. A solution to work around the phenomenon to describe is just to keep a table of variables that cannot be instantiated to SProp (i.e. we quotient the quality graph by the inclusion Prop ⊆ Type) as a second unification state. For now given the little amount of broken code I don't want to disrupt the unification algorithm too much, but it will become possible with the current infrastructure (and it won't even break the API).